File: story.js
// WORK IN PROGRESS!

var STRING_LITERAL_CHARS = ['\'', '\"'];
var WHITESPACE = [' ', '\t'];
// strings of symbols have implicit whitespace surrounding them
var SYMBOLS = ['*', '+', '-', '/', '<', '=', '>', '^'];
var INVALID_CHARS = ['!', '#', '$', '%', '&', '(', ')', ',', ';', '?', '@',
                     '\\', '`', '{', '|', '}', '~'];


if(typeof(String.prototype.trim) === "undefined") {
    String.prototype.trim = function() {
        return String(this).replace(/^\s+|\s+$/g, '');
    };
}


// assumes that the expression has no errors
function compileExpression(expression, language) {
    expression = _removeComments(expression);
    return _compileRecursive(expression, language);
}

function checkExpression(expression, language) {
    var errors = [ ]
    expression = _removeComments(expression);
    
    var lastBracketDepth = 0;
    var lastLine = 0;
    var lastOpenBracketLine = 0;
    var lastInString = false;
    _iterateExpression(expression,
    function(c, i, line, inString, bracketDepth) {
        if(inString) {
            if(c == '\n') {
                errors.push(incompleteStringError(line));
            }
        } else {
            if(INVALID_CHARS.indexOf(c) != -1) {
                errors.push(invalidCharacterError(c, line));
            }
            if(c == ']' && bracketDepth < 0) {
                errors.push(unmatchedCloseBracketError(line));
            }
            if(c == '[')
                lastOpenBracketLine = line;
        }
        lastBracketDepth = bracketDepth;
        if(WHITESPACE.indexOf(c) == -1 && c != '\n')
            lastLine = line;
        lastInString = inString;
    });
    if(lastBracketDepth > 0) {
        errors.push(unmatchedOpenBracketError(lastOpenBracketLine));
    }
    if(lastInString) {
        errors.push(incompleteStringError(lastLine));
    }
    
    if(errors.length != 0)
        return errors;
    
    // recursive expression scan
    var type = _checkRecursive(expression, language, 0, errors);
    if(type != "Nothing")
        errors.push(typeNotNothingError(lastLine));
    
    return errors;
}


// returns the type name
function _checkRecursive(expression, language, startLine, errorList) {
    expression = expression.trim();
    if(expression.length == 0) {
        errorList.push(emptyExpressionError(startLine));
        return "Nothing";
    }
    
    expressionType = _detectExpressionType(expression);
    
    if(expressionType == "number") {
        return "Number";
    } else if(expressionType == "multiline") {
        var prevLineIndex = 0;
        var types = [];
        var typeLines = [];
        _iterateExpression(expression + "\n",
        function(c, i, line, inString, bracketDepth) {
            if(c == '\n' && bracketDepth == 0) {
                var lineExpr = expression.substring(prevLineIndex, i).trim();
                if(lineExpr.length != 0) {
                    var type =
                        _checkRecursive(lineExpr, language, line, errorList);
                    types.push(type);
                    typeLines.push(line);
                }
                prevLineIndex = i + 1;
            }
        });
        
        for(var i = 0; i < types.length - 1; i++) {
            if(types[i] != "Nothing")
                errorList.push(typeNotNothingError(typeLines[i]));
        }
        return types[types.length - 1];
    } else if(expressionType == "blank") {
        errorList.push(blankError(startLine));
        return "Anything";
    } else if(expressionType == "text") {
        return "Text";
    } else if(expressionType == "command") {
        return "Nothing"; // for testing
    } else {
        errorList.push(unknownError(startLine));
    }
}


function _compileRecursive(expression, language) {
    expression = expression.trim();
    var len = expression.length;
    if(len == 0)
        return null;
    
    expressionType = _detectExpressionType(expression);
    
    if(expressionType == "number") {
        // number literal
        return {
            "literalText": expression
        };
    } else if(expressionType == "multiline") {
        var lines = [];
        var prevLineIndex = 0;
        _iterateExpression(expression,
        function(c, i, line, inString, bracketDepth) {
            if(c == '\n' && bracketDepth == 0) {
                var line = expression.substring(prevLineIndex, i).trim();
                if(line.length != 0)
                    lines.push(line);
                prevLineIndex = i + 1;
            }
        });
        var line = expression.substring(prevLineIndex, len).trim();
        if(line.length != 0)
            lines.push(line);
        var compiledLines = [];
        for(var i = 0, numLines = lines.length; i < numLines; i++) {
            compiledLines.push(_compileRecursive(lines[i], language));
        }
        return {
            "Lines": compiledLines
        };
    } else if(expressionType == "blank") {
        throw "You must remove all blanks (underscores) before compiling!";
    } else if(expressionType == "text") {
        // text literal
        return {
            "literalText": expression.substring(1, expression.length - 1)
        };
    } else if(expressionType == "command") {
        var parts = [];
        var prevSpaceIndex = 0, symbolString = false, wasInString = false;
        expression = expression + " "; // process last part
        _iterateExpression(expression,
        function(c, i, line, inString, bracketDepth) {
            var isSymbol = SYMBOLS.indexOf(c) != -1;
            var isWhitespace = WHITESPACE.indexOf(c) != -1;
            var isStringChar = STRING_LITERAL_CHARS.indexOf(c) != -1;
            if((bracketDepth == 0 || (bracketDepth == 1 && c == '['))
                    && (!inString || isStringChar)) {
                if(isWhitespace || c == '[' || (isSymbol && !symbolString)
                        || (!isSymbol && symbolString)
                        || (!(inString || isStringChar) && wasInString)
                        || ((inString || isStringChar) && !wasInString)) {
                    symbolString = isSymbol;
                    var part = expression.substring(prevSpaceIndex, i).trim();
                    if(part.length != 0) {
                        if(part.charAt(0) == '['
                                && part.charAt(part.length - 1) == ']') {
                            part = part.substring(1, part.length - 1).trim();
                            // [ ] is still allowed
                        }
                        parts.push(part);
                    }
                    prevSpaceIndex = i;
                }
            } else {
                symbolString = false;
            }
            wasInString = inString || isStringChar;
        });
        
        // determine which command this is...
        var matchingCommand = null;
        for(var i = 0, numCommands = language.Commands.length;
                i < numCommands; i++) {
            var command = language.Commands[i];
            var keywordWords = command.BasePhrase.keyword.split(" ");
            var hasPrefix = command.PrecedingArgument !== undefined;
            var minimumLength = keywordWords.length;
            if(hasPrefix)
                minimumLength += 1;
            if(parts.length < minimumLength)
                continue;
            var matches = true;
            for(var partIndex = 0; partIndex < keywordWords.length;
                    partIndex++) {
                if(parts[partIndex + (hasPrefix ? 1 : 0)]
                        != keywordWords[partIndex]) {
                    matches = false;
                    break;
                }
            }
            
            if(matches) {
                matchingCommand = command;
                break;
            }
        }
        if(matchingCommand == null) {
            throw "No matching command for " + expression;
        }
        var compiledCommand = {
            "keyword": matchingCommand.BasePhrase.keyword,
            "Arguments": [ ],
            "optionalPhraseKeywords": [ ]
        }
        
        var partIndex = 0;
        if(matchingCommand.PrecedingArgument !== undefined) {
            compiledCommand.PrecedingExpression =
                _partValue(parts[0], matchingCommand.PrecedingArgument,
                    language);
            partIndex = 1;
        }
        // skip keywords
        partIndex += command.BasePhrase.keyword.split(" ").length;
        for(var phrasePartIndex = 0,
                numPhraseParts = matchingCommand.BasePhrase.Parts.length;
                phrasePartIndex < numPhraseParts; phrasePartIndex++) {
            var part = parts[partIndex];
            var phrasePart = matchingCommand.BasePhrase.Parts[phrasePartIndex];
            if(phrasePart.id !== undefined) {
                compiledCommand.Arguments.push({
                    "id": phrasePart.id,
                    "Expression": _partValue(part, phrasePart, language)
                });
            }
            partIndex++;
        }
        // optional phrases...
        
        return compiledCommand
    }
}


function _detectExpressionType(expression) {
    if(!isNaN(expression)) {
        return "number"
    }
    
    var textInLine = false, textInPreviousLine = false;
    var multiline = false, textLiteral = true, underscores = true;
    _iterateExpression(expression,
    function(c, i, line, inString, bracketDepth) {
        if(c != '_')
            underscores = false;
        
        if(WHITESPACE.indexOf(c) != -1) {
            if(!inString)
                textLiteral = false;
            return;
        }
        
        if(c == '\n' && bracketDepth == 0) {
            textInPreviousLine = textInLine;
            textInLine = false;
            textLiteral = false;
            return;
        }
        
        textInLine = true;
        if(textInPreviousLine) {
            multiline = true;
            return false;
        }
        
        if(!inString && STRING_LITERAL_CHARS.indexOf(c) == -1)
            textLiteral = false;
    });
    
    if(multiline) {
        return "multiline";
    } else if(underscores) {
        return "blank";
    } else if(textLiteral) {
        return "text";
    } else {
        return "command";
    }
}


function _partValue(part, partDefinition, language) {
    if(partDefinition.type == "") {
        return {
            "literalText": part
        };
    } else {
        return _compileRecursive(part, language);
    }
}


function _removeComments(expression) {
    var lines = expression.split("\n");
    for(var i = 0, numLines = lines.length; i < numLines; i++) {
        var line = lines[i];
        _iterateExpression(line,
        function(c, j, line, inString, bracketDepth) {
            if(!inString && c == ':') {
                line = line.substring(0, j);
                return false;
            }
        });
        
        lines[i] = line;
    }
    
    return lines.join("\n");
}


// f(char c, int i, int line, bool inString, int bracketDepth)
// return false to stop
function _iterateExpression(expression, f) {
    var inString = false, stringStartChar, bracketDepth = 0,
        line = 0, newLine = false;
    for(var i = 0, len = expression.length; i < len; i++) {
        var c = expression.charAt(i);
        
        if(newLine) {
            newLine = false;
            line++;
            inString = false;
        }
        if(c == '\n')
            newLine = true;
        
        if(inString) {
            if(c == stringStartChar) {
                inString = false;
            }
        } else {
            if(STRING_LITERAL_CHARS.indexOf(c) != -1) {
                inString = true;
                stringStartChar = c;
            }
            
            if(c == '[')
                bracketDepth += 1;
            if(c == ']')
                bracketDepth -= 1;
        }
        
        if(f(c, i, line, inString, bracketDepth) === false)
            break;
    }
}


function ScriptError(message, type, line) {
    this.message = message;
    this.type = type;
    this.line = line;
}

function unknownError(line) {
    return new ScriptError(
        "An unknown error occurred while interpreting the expression", "error",
        line);
}

function blankError(line) {
    return new ScriptError("Fill in the blank", "blank", line);
}

function invalidCharacterError(character, line) {
    return new ScriptError("Invalid character " + character, "error", line);
}

function incompleteStringError(line) {
    return new ScriptError("Text is missing an ending quote", "error", line);
}

function unmatchedOpenBracketError(line) {
    return new ScriptError(
        "Opening bracket doesn't have a matching closing bracket", "error",
        line);
}

function unmatchedCloseBracketError(line) {
    return new ScriptError(
        "Closing bracket doesn't have a matching opening bracket", "error",
        line);
}

function typeNotNothingError(line) {
    return new ScriptError("Incomplete statement", "error", line);
}

function emptyExpressionError(line) {
    return new ScriptError("Empty expression", "error", line);
}

File: story.py

### Syntax Constants ###

INVALID_CHARS = ['!', '#', '$', '%', '&', '(', ')', ',', ';', '?', '@', '\\',
                 '`', '{', '|', '}', '~']
WHITESPACE = [' ', '\t']
# strings of symbols have implicit whitespace surrounding them
SYMBOLS = ['*', '+', '-', '/', '<', '=', '>', '^']
TEXT_LITERAL_CHARS = ["'", '"']
COMMENT_CHAR = ':'
OPEN_BRACKET = '['
CLOSE_BRACKET = ']'
BLANK = '_'
MINUS = '-'
# not including minus
NUMBER_LITERAL_CHARS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.']
# not including letters and numbers
WORD_CHARS = ['_']


### Errors, warnings, messages ###

class CompileMessage:

    def __init__(self, message, messageType, location, length):
        self.message = message
        self.type = messageType
        self.location = location
        self.length = length

    def __repr__(self):
        return self.type + " at " + str(self.location) + " - "\
               + str(self.location + self.length) + ": " + self.message

class CompileError(CompileMessage):

    def __init__(self, message, location, length):
        super().__init__(message, "error", location, length)

class CompileWarning(CompileMessage):

    def __init__(self, message, location, length):
        super().__init__(message, "warning", location, length)


class BlankMessage(CompileMessage):

    def __init__(self, location, length):
        super().__init__("Fill in the blank", "blank", location, length)

class UnknownError(CompileError):

    def __init__(self, location, length):
        super().__init__("An unknown error occurred", location, length)

class InvalidCharacterError(CompileError):

    def __init__(self, character, location):
        super().__init__("Invalid character: " + character, location, length=1)

class IncompleteTextError(CompileError):

    def __init__(self, location):
        super().__init__("Text is missing an ending quote", location, length=1)

class UnmatchedOpenBracketError(CompileError):

    def __init__(self, location):
        super().__init__("Opening bracket doesn't have a matching closing "
                         "bracket", location, length=1)

class UnmatchedCloseBracketError(CompileError):

    def __init__(self, location):
        super().__init__("Closing bracket doesn't have a matching opening "
                         "bracket", location, length=1)

class InvalidNumberLiteralError(CompileError):

    def __init__(self, location, length):
        super().__init__("Badly formatted number", location, length)

class TypeNotNothingError(CompileError):

    def __init__(self, location, length):
        super().__init__("Incomplete statement", location, length)

class EmptyExpressionError(CompileError):

    def __init__(self, location, length):
        super().__init__("Empty expression", location, length)


### Tokenizer ###

class Token:

    def __init__(self, location, text=''):
        self.text = text
        self.location = location

    def __repr__(self):
        return repr(self.text) + " @ " + str(self.location)

# check token type:
# these assume a correctly formatted token, which the tokenize function should
# produce as long as there are no errors

def tokenIsNewline(text):
    return text == '\n'

def tokenIsOpenBracket(text):
    return text == OPEN_BRACKET

def tokenIsCloseBracket(text):
    return text == CLOSE_BRACKET

def tokenIsWord(text):
    if text == '':
        return False
    return text[0] != BLANK and (text[0].isalpha() or text[0] in WORD_CHARS)

def tokenIsText(text):
    if text == '':
        return False
    return text[0] in TEXT_LITERAL_CHARS

def tokenIsSymbol(text):
    if text == '':
        return False
    if text[0] == MINUS:
        # make sure this isn't a number
        if len(text) == 1:
            return True
        else:
            return text[1] in SYMBOLS
    else:
        return text[0] in SYMBOLS

def tokenIsNumber(text):
    if text == '':
        return False
    elif len(text) == 1:
        return text in NUMBER_LITERAL_CHARS
    elif text[0] == MINUS:
        return text[1] in NUMBER_LITERAL_CHARS
    else:
        return text[0] in NUMBER_LITERAL_CHARS

def tokenIsBlank(text):
    if text == '':
        return False
    return text[0] == BLANK

# return a tuple: (list of Tokens, list of CompileMessages)
def tokenize(script):
    script += '\n'

    messages = []
    tokens = []

    currentToken = Token(0)
    # identify currentToken type:
    symbol = False
    numberLiteral = False
    blank = False
    word = False

    inComment = False
    inText = False
    textStartChar = None
    bracketDepth = 0
    lineIsEmpty = True

    lastOpenBracketI = 0

    for i in range(0, len(script)):
        c = script[i]

        if inComment:
            if c == '\n':
                inComment = False
            else:
                continue

        if inText:
            currentToken.text += c
            if c == textStartChar or c == '\n':
                if c == '\n':
                    messages.append(IncompleteTextError(i))
                    currentToken.text = currentToken.text[:-1]
                inText = False
                tokens.append(currentToken)
                currentToken = None
            continue

        if c in INVALID_CHARS:
            messages.append(InvalidCharacterError(c, i))

        if c != '\n' and not c in WHITESPACE:
            lineIsEmpty = False


        # check for end of token (no continues here):

        if symbol and not c in SYMBOLS:
            tokens.append(currentToken)
            currentToken = None
            symbol = False

        if numberLiteral and not c in NUMBER_LITERAL_CHARS:
            tokens.append(currentToken)
            try:
                n = float(currentToken.text)
            except ValueError:
                messages.append(InvalidNumberLiteralError(currentToken.location,
                                                        len(currentToken.text)))
            currentToken = None
            numberLiteral = False

        if blank and not c == BLANK:
            tokens.append(currentToken)
            messages.append(BlankMessage(currentToken.location,
                                         len(currentToken.text)))
            currentToken = None
            blank = False

        if word and (not c.isdigit()) and (not c.isalpha()) \
                and not c in WORD_CHARS:
            tokens.append(currentToken)
            currentToken = None
            word = False

        # text / comments are special

        if c in TEXT_LITERAL_CHARS:
            inText = True
            textStartChar = c
            currentToken = Token(i, c)
            continue

        if c == COMMENT_CHAR:
            inComment = True
            continue

        # non-token checks

        if c == '\n':
            if not lineIsEmpty:
                tokens.append(Token(i, '\n'))
            lineIsEmpty = True

        if c == OPEN_BRACKET:
            bracketDepth += 1
            tokens.append(Token(i, c))
            lastOpenBracketI = i

        if c == CLOSE_BRACKET:
            if bracketDepth == 0:
                messages.append(UnmatchedCloseBracketError(i))
            else:
                bracketDepth -= 1
            tokens.append(Token(i, c))

        # check for start/continuation of token:

        if symbol or numberLiteral or blank or word:
            currentToken.text += c
            # can't have multiple tokens at once
            continue

        if c in SYMBOLS:
            symbol = True
            currentToken = Token(i, c)

        elif c in NUMBER_LITERAL_CHARS:
            numberLiteral = True
            if len(tokens) > 0 and tokens[-1].text == MINUS \
                    and _minusIsNegative(tokens):
                minusToken = tokens.pop()
                currentToken = Token(minusToken.location, MINUS + c)
            else:
                currentToken = Token(i, c)

        elif c == BLANK:
            blank = True
            currentToken = Token(i, c)

        elif c in WORD_CHARS or c.isalpha():
            word = True
            currentToken = Token(i, c)

    # end for i in range(0, len(script)):

    if bracketDepth > 0:
        messages.append(UnmatchedOpenBracketError(lastOpenBracketI))

    return tokens, messages
# end tokenize

# Used for when the most recent token is a minus sign and a number is about
# to be added. Should the minus sign be interpreted as the negative sign of the
# number? (True) or as a subtract symbol? (False)
# There shouldn't be any ambiguity in these rules.
def _minusIsNegative(tokens):
    if len(tokens) == 1:
        # nothing before it, so nothing to subtract from.
        return True

    beforeMinus = tokens[-2]
    if tokenIsOpenBracket(beforeMinus.text) \
            or tokenIsSymbol(beforeMinus.text) \
            or tokenIsWord(beforeMinus.text) \
            or tokenIsNewline(beforeMinus.text):
        # open bracket or newline: start of command, nothing to subtract from
        # word or symbol: in the middle of a command that isn't a subtraction
        return True

    # beforeMinus is a number, blank, close bracket, or string

    if len(tokens) == 2:
        # the number, blank, or string is the first argument of a subtraction
        return False

    if tokenIsCloseBracket(beforeMinus.text):
        # find token before matching open bracket
        i = len(tokens) - 3
        searchBracketDepth = 1
        while i >= 0:
            t = tokens[i]
            if tokenIsOpenBracket(t.text):
                searchBracketDepth -= 1
            elif tokenIsCloseBracket(t.text):
                searchBracketDepth += 1
            if searchBracketDepth == 0:
                break
            i -= 1
        if i <= 0:
            return False
        beforeBeforeMinus = tokens[i - 1]
    else:
        beforeBeforeMinus = tokens[-3]

    if tokenIsOpenBracket(beforeBeforeMinus.text) \
            or tokenIsNewline(beforeBeforeMinus.text):
        # this is a subtraction command
        return False

    return True
